lib/instrumentation/modules/hapi.js

```diff
@@ -14,113 +14,137 @@ module.exports = function (hapi, agent, version) {
 
   debug('shimming hapi.Server.prototype.initialize')
 
-  shimmer.wrap(hapi.Server.prototype, 'initialize', function (orig) {
-    return function () {
-      // Hooks that are always allowed
-      if (typeof this.on === 'function') {
-        this.on('request-error', function (req, err) {
-          agent.captureError(err, {request: req.raw && req.raw.req})
-        })
-
-        this.on('log', function (event, tags) {
-          if (!event || !tags.error) return
-
-          var payload = {custom: event} // TODO: Find better location to put this than custom
-
-          var err = event.data
-          if (!(err instanceof Error) && typeof err !== 'string') {
-            err = 'hapi server emitted a log event tagged error'
-          }
-
-          agent.captureError(err, payload)
-        })
-
-        this.on('request', function (req, event, tags) {
-          if (!event || !tags.error) return
-
-          var payload = {
-            custom: event, // TODO: Find better location to put this than custom
-            request: req.raw && req.raw.req
-          }
-
-          var err = event.data
-          if (!(err instanceof Error) && typeof err !== 'string') {
-            err = 'hapi server emitted a request event tagged error'
-          }
-
-          agent.captureError(err, payload)
-        })
-      } else {
-        debug('unable to enable hapi error tracking')
+  if (semver.satisfies(version, '>=17')) {
+    shimmer.massWrap(hapi, ['Server', 'server'], function (orig) {
+      return function (options) {
+        var res = orig.apply(this, arguments)
+        patchServer(res)
+        return res
       }
-
-      // When the hapi server has no connections we don't make connection
-      // lifecycle hooks
-      if (this.connections.length === 0) {
-        debug('unable to enable hapi instrumentation on connectionless server')
+    })
+  } else {
+    shimmer.wrap(hapi.Server.prototype, 'initialize', function (orig) {
+      return function () {
+        patchServer(this)
         return orig.apply(this, arguments)
       }
+    })
+  }
+
+  function patchServer (server) {
+    // Hooks that are always allowed
+    if (typeof server.on === 'function') {
+      attachEvents(server)
+    } else if (typeof server.events.on === 'function') {
+      attachEvents(server.events)
+    } else {
+      debug('unable to enable hapi error tracking')
+    }
+
+    // When the hapi server has no connections we don't make connection
+    // lifecycle hooks
+    var conns = server.connections
+    if (conns && conns.length === 0) {
+      debug('unable to enable hapi instrumentation on connectionless server')
+      return
+    }
+
+    // Hooks that are only allowed when the hapi server has connections
```

Comment:
Not a 100% correct statement anymore

Reply:
The check is still fine afaik

Reply:
@watson Any thoughts on wording here?

Reply:
@AdriVanHoudt Do you mean related to the removal of multi-connections support in hapi 17?
 What would be a more appropriate comment?

Reply:
As far as I can tell hapi 17 will have it always.
In 16 it only becomes available after adding a connecting but in 17 you can't run a server without it. The moment you create the server the ext will be there in 17.
So maybe `When we can hook into the lifecycle to do proper transaction naming` or something?

Reply:
@AdriVanHoudt how about just `Hooks that are only allowed when the hapi server has connections (with hapi 17+ this is always the case)`

Reply:
updated.